cola/widgets/startup.py

```diff
@@ -265,16 +336,397 @@ def handle_broken_repo(self, index):
+
+        if self._is_bookmark_mode():
+            rename = settings.rename_bookmark
+        elif self._is_recent_mode():
+            rename = settings.rename_recent
+        else:
+            rename = _disable_rename
+
+        if rename(item.path, item.name, new_name):
+            settings.save()
+            item.name = new_name
+
+    def apply_fn(self, fn, *args, **kwargs):
+        item = self.selected_item()
+        if item:
+            fn(item, *args, **kwargs)
+
+    def copy(self):
+        self.apply_fn(lambda item: qtutils.set_clipboard(item.path))
+
+    def open_default(self):
+        context = self.context
+        self.apply_fn(lambda item: cmds.do(cmds.OpenDefaultApp, context, [item.path]))
+
+    def set_default_repo(self):
+        self.apply_fn(self.set_default_item)
+
+    def set_default_item(self, item):
+        context = self.context
+        cmds.do(cmds.SetDefaultRepo, context, item.path)
+        self.refresh()
+
+    def clear_default_repo(self):
+        self.apply_fn(self.clear_default_item)
+
+    def clear_default_item(self, _item):
+        context = self.context
+        cmds.do(cmds.SetDefaultRepo, context, None)
+        self.refresh()
+
+    def rename_repo(self):
+        index = self.currentIndex()
+        self.edit(index)
+
+    def accept_repo(self):
+        self.apply_fn(self.accept_item)
+
+    def accept_item(self, _item):
+        if self.state() & self.EditingState:
+            current_index = self.currentIndex()
+            widget = self.indexWidget(current_index)
+            if widget:
+                self.commitData(widget)
+            self.closePersistentEditor(current_index)
+            self.refresh()
+        else:
+            self.open_selected_repo()
+
+    def open_new_repo(self):
+        context = self.context
+        self.apply_fn(lambda item: cmds.do(cmds.OpenNewRepo, context, item.path))
+
+    def launch_editor(self):
+        context = self.context
+        self.apply_fn(lambda item: cmds.do(cmds.Edit, context, [item.path]))
+
+    def launch_terminal(self):
+        context = self.context
+        self.apply_fn(lambda item: cmds.do(cmds.LaunchTerminal, context, item.path))
+
+    def delete_recent(self):
+        """Removes a selected repo item from the recent list"""
+        cmd = cmds.RemoveRecent
+        self.delete(cmd)
+
+    def delete_bookmark(self):
+        """Removes a selected repo item from the bookmark list"""
+        cmd = cmds.RemoveBookmark
+        self.delete(cmd)
+
+    def delete(self, cmd):
+        item = self.selected_item()
+        context = self.context
+        if not item:
+            return
+        ok, _, _, _ = cmds.do(cmd, context, item.path, item.name, icon=icons.discard())
+        if ok:
+            self.refresh()
+
+    def remove_missing(self):
+        """Remove missing entries from the favorites/recent file list"""
+        settings = self.context.settings
+        settings.remove_missing_recent()
```

Comment:
OK! I'll do it.

Reply:
About `remove_missing` -- I think this action doesn't even need to be item-specific. I think we would want to remove missing bookmarks and missing recent entries all in a single call here.

For what it's worth, the remove-missing context menu entry seems like it should be present even when right-clicking in the widget in the blank area without any icons. Not important for now, but just noting in case it helps. This should be the only UI action that will modify both the bookmarks and recents from a single action. The rest can limit their scope to bookmarks when `is_bookmark` is True.